### ISSUE TITLE
Update/Add linux-modules-extra-azure to images by default

### ISSUE THREAD

**Issue by bluca** [actions/runner-images#7587]
### Tool name

linux-modules-extra-azure

### Tool license

Same as the kernel image

### Add or update?

- [X] Add
- [ ] Update

### Desired version

Same version as linux-image-azure

### Approximate size

125.93 MB

### Brief description of tool

This package contains extra kernel modules for the linux-image-azure kernel package that the runner images are using. We're looking to mount erofs filesystems for [systemd](https://github.com/systemd/systemd) integration tests which need the erofs kernel module which is available from linux-modules-extra-azure.

We need this to be installed in the regular image because the version of the modules needs to match the version of the running kernel, otherwise the modules can't be loaded by the running kernel. We can't install the version that matches the running kernel because newer versions get released and older versions of the package are removed from the Ubuntu repositories.

Another option is to make sure that a version of the linux-modules-extra-azure package that matches the kernel package that is shipped in the current image version is always available.

Currently, if I do sudo apt-get install linux-modules-extra-$(uname -r), the command fails because the matching version is not available in the mirrors. This requires the GA team to make sure the version is available in the azure apt mirrors until the runner image with the same kernel version is no longer in use.

These kernel modules cannot lag behind the actual kernel update when you install both together, as they are all built from the same source package. Canonical choose to split it out in multiple binary packages, for some reasons. When you install the kernel image package and that extra-module package at the same time, they are guaranteed to be compatible, as the unversioned metapackages that you are referencing point to the exact same kernel build. This is guaranteed by how the packages are set up.

The problem arises because the kernel package is installed whenever the image is built, and we are trying to install the extra modules package long after that has happened, on the user script. By installing both at image creation time, they will be in sync.

You can see that linux-image-azure and linux-modules-extra-azure are part of the same source package and thus are built and published together here:

https://packages.ubuntu.com/source/jammy-updates/linux-meta-azure

And if for some weird, strange circumstance that does go out of sync at image creation time, nothing bad happens, there's simply some unusable kernel modules. The installation location is named after its ABI, so it won't affect the actual kernel.

To be extra clear: these are not out-of-tree or 3rd party kernel modules. They have been part of the mainline kernel since ages, and are all built together with the kernel image itself. It's just that Canonical choose to organize them in separate packages for some reason, built at the same time. They cannot go out of sync or be incompatible.

### URL for tool's homepage

https://packages.ubuntu.com/source/jammy-updates/linux-meta-azure

### Platforms where you need the tool

- [ ] Azure DevOps
- [X] GitHub Actions

### Runner images where you need the tool

- [ ] Ubuntu 20.04
- [X] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [ ] Windows Server 2019
- [ ] Windows Server 2022

**Comment by YT-Advanced** [comment]
@bluca can you help me with sudo mount option for EROFS ?

**Comment by YT-Advanced** [comment]
I need it too

**Comment by erik-bershel** [comment]
Hey @bluca! We'll think about adding it.

**Comment by sergei-pyshnoi** [comment]
Hello @bluca . We have some thoughts on adding these modules to the image. It is impossible to predict the system behavior if will be attempt upgrade all packages using apt upgrade. In this case version of linux-modules-extra-azure can be higher (if those available) than the version of the kernel. For the most pipelines/workflows it is regular practice to upgrading existing packages. With this,  it will high cost for maintaining solution for keep the identical version of modules and kernel.

**Comment by bluca** [comment]
> Hello @bluca . We have some thoughts on adding these modules to the image. It is impossible to predict the system behavior if will be attempt upgrade all packages using apt upgrade. In this case version of linux-modules-extra-azure can be higher (if those available) than the version of the kernel. For the most pipelines/workflows it is regular practice to upgrading existing packages. With this, it will high cost for maintaining solution for keep the identical version of modules and kernel.

Uh? But it doesn't matter, because the machine cannot be rebooted anyway by the user, so even if upgrades happen it won't have any effect, as the new kernel (and thus the new modules) will not be in use, so what's the issue exactly?

**Comment by sergei-pyshnoi** [comment]
> > Hello @bluca . We have some thoughts on adding these modules to the image. It is impossible to predict the system behavior if will be attempt upgrade all packages using apt upgrade. In this case version of linux-modules-extra-azure can be higher (if those available) than the version of the kernel. For the most pipelines/workflows it is regular practice to upgrading existing packages. With this, it will high cost for maintaining solution for keep the identical version of modules and kernel.
> 
> Uh? But it doesn't matter, because the machine cannot be rebooted anyway by the user, so even if upgrades happen it won't have any effect, as the new kernel (and thus the new modules) will not be in use, so what's the issue exactly?

By the way, can you please provide exact list of modules that you need in you're tests?

**Comment by bluca** [comment]
For now dmcrypt, dmverity, squashfs, btrfs, erofs, probably a few more. There will probably be more in the future. Can you please reopen the PR?

**Comment by mripard** [comment]
Hi, I'm interested in this as well.

> In this case version of linux-modules-extra-azure can be higher (if those available) than the version of the kernel. For the most pipelines/workflows it is regular practice to upgrading existing packages. With this, it will high cost for maintaining solution for keep the identical version of modules and kernel.

Even though that can definitely happen, it's not something to be concerned about. `modprobe` will only load modules installed for the same version than the kernel currently running (so it won't use the new version modules) and if someone was to try to load one of those modules by hand, the kernel has checks to prevent us from loading a module built from a different version or a different configuration than what is currently running.

So in short, yes, it can happen, but it won't affect the system in any way.

> By the way, can you please provide exact list of modules that you need in you're tests?

I would need vkms.

Thanks!

**Comment by fnordahl** [comment]
I'm interested in this as well, in this instance I need the `vrf` kernel module.

In the mean time, something like the below step does provide it for me even if the image is running a kernel older than the most recent available version.

```yaml
    - name: system-level-dependencies
      run: |
        sudo apt update
        sudo apt -y install linux-modules-extra-$(uname -r)
```

As the OP points out, it would be nicer if this package was just included in the image.
